### PR TITLE
HPCC-29614 Preserve ECM variable types in generated service XML during build

### PR DESCRIPTION
**Pull request by kenrowland**
Added HIDL compatible mode to esdl compiler and esdl-xml command that outputs an attribute specifying the type that matches HIDL generated code output.

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: tools/esdlcmd-xml/esdl2xml.cpp [hunk 1]
```diff
@@ -73,6 +74,10 @@ void parseCommandLine(int argc, char* argv[], Esdl2Esxdl * cmd)
             {
                 cmd->setVerbose(true);
             }
+            else if (stricmp(argv[i], "-m")==0 || stricmp(argv[i], "--hidl-mode")==0)
+            {
+                cmd->setHidlMode(true);
```
Comment (afishbeck) [inline tools/esdlcmd-xml/esdl2xml.cpp:79]: Again, no need for an option when just generating the xml.

File: tools/esdlcomp/esdlcomp.h [hunk 2]
```diff
@@ -605,6 +605,12 @@ class ParamInfo
         {
             out.appendf("\t\t<EsdlElement name='%s'", name);
             toStringXmlAttr(out);
+
+            if (xsdtype && *xsdtype && hidlMode)
+            {
+                out.appendf(" hidlType='%s'", xsdtype);
```
Comment (afishbeck) [inline tools/esdlcomp/esdlcomp.h:611]: Looking at esxdl I see it uses underscores not camel case.  hidl_type would match other attributes.
But now that I see how this works I think an even better name would be xsd_extended_type.  That's what we are really outputting.  Hidl xsd compatibility really means "using the xsd extended type".

File: tools/esdlcomp/esdlcomp.h [hunk 3]
```diff
@@ -605,6 +605,12 @@ class ParamInfo
         {
             out.appendf("\t\t<EsdlElement name='%s'", name);
             toStringXmlAttr(out);
+
+            if (xsdtype && *xsdtype && hidlMode)
```
Comment (afishbeck) [inline tools/esdlcomp/esdlcomp.h:609]: xsdtype is set for general strings that don't use an extended type.  This code results in:

		<EsdlElement name='PriorityDesc' type='string' hidlType='string'/>
		<EsdlElement name='PriorityLevel' type='int'/>
		<EsdlElement name='DateTimeScheduled' type='string' hidlType='dateTime'/>

Notice that the string type has both type and hidl_type equal to "string".

You could just check for that here.

if (xsdtype && *xsdtype && strcmp(xsdtype, "string")!=0)


Combining that with my other comments the output would look like:

		<EsdlElement name='PriorityDesc' type='string'/>
		<EsdlElement name='PriorityLevel' type='int'/>
		<EsdlElement name='DateTimeScheduled' type='string' xsd_extended_type='dateTime'/>

**Review by afishbeck — CHANGES_REQUESTED** [review]
@kenrowland Some comments.

**Review by asselitx — COMMENTED** [review]
I don't have anything to add beyond Tony's comments.

File: tools/esdlcmd-xml/esdl2xml.cpp (hunk 1 above)
Comment (kenrowland) [reply tools/esdlcmd-xml/esdl2xml.cpp]: Kept the option, but now only used to turn it off if necessary.

**Review by afishbeck — COMMENTED** [review]
@kenrowland some comments.

**Review by asselitx — APPROVED** [review]

**Review by afishbeck — APPROVED** [review]

**Comment by kenrowland** [comment]
@ghalliday  Please merge

File: tools/esdlcmd-xml/esdl2xml.hpp [hunk 4]
```diff
@@ -43,6 +43,9 @@ class Esdl2Esxdl : public CInterface
     void setVerbose(bool verbose){optVerbose = verbose;};
     bool getVerbose(){return optVerbose;};
 
+    void setNoExtendedAttributes(bool mode){ optNoExtendedAttributes = mode;};
+    bool getNoExtendedAttributes(){return optNoExtendedAttributes;};
```
Comment (ghalliday) [inline tools/esdlcmd-xml/esdl2xml.hpp:47]: could be a const function

File: tools/esdlcomp/esdlcomp.cpp [hunk 5]
```diff
@@ -1340,5 +1409,16 @@ void ESDLcompiler::write_esxdl()
     }
 }
 
+const void ESDLcompiler::setExtendedAttributes(bool mode)
+{
+    extendedAttributes = mode;
+}
+
+const bool ESDLcompiler::getExtendedAttributes()
```
Comment (ghalliday) [inline tools/esdlcomp/esdlcomp.cpp:1417]: could be const

**Review by ghalliday — APPROVED** [review]

**Pull request merged**